zjit/src/hir.rs

```diff
@@ -1199,6 +1203,44 @@ impl Function {
         }
     }
 
+    fn try_rewrite_freeze(&mut self, block: BlockId, orig_insn_id: InsnId, self_val: InsnId) {
+        let self_type = self.type_of(self_val);
+
+        let redefined_op_flag = if self_type.is_subtype(types::StringExact) {
+            Some(STRING_REDEFINED_OP_FLAG)
+        } else if self_type.is_subtype(types::ArrayExact) {
+            Some(ARRAY_REDEFINED_OP_FLAG)
+        } else if self_type.is_subtype(types::HashExact) {
+            Some(HASH_REDEFINED_OP_FLAG)
+        } else {
+            None
+        };
+
+        if let Some(obj) = self_type.ruby_object() {
```

Comment:
Does this triple `if` look weird? Maybe it's better if I just write out the full `push_insn` 3 times? Let me know if you think I should keep playing around with it.

Reply:
It's a little awkward but I think it's fine for now. Means our APIs are a little awkward. We can think about it and come back to this later since it's a small change

Reply:
After taking a break I found something I like a little better.